### PR TITLE
1.16

### PR DESCRIPTION
**Pull request by extremeheat**
WIP, opening to keep track of progress/discussion

* [x] Raknet
* [x] JWT verification
* [x] Encryption
* [x] NBT - https://github.com/PrismarineJS/prismarine-nbt/pull/35 done at extremeheat/prismarine-nbt#le
* [ ] LevelDB chunks - https://github.com/extremeheat/bedrock-provider/issues/1 (TODO: Entities, tiles)
* [ ] Docs - protocol docs, started work directly in the [protocol yaml](https://github.com/extremeheat/pocket-minecraft-protocol/blob/new/data/new/proto.yml#L19), todo: generate typescript definitions
* [ ] ProtoDef - compiler needs some changes to getField to allow for reserved names + 64-bit unsigned varint (non-zigzag)
* [x] Biomes, creative content NBT
* [ ] Documentation

- [x] Server
  - [x] Block palette 
  - [x] Accepts minet clients
  - [x] Accepts vanilla clients
  - [x] Spawns - need chunk implementation to spawn chunks
- [x] Client
  - [x] Authenticates with xbox/mc nearly same as https://github.com/PrismarineJS/node-minecraft-protocol/pull/806 - possibly could be separated to own package ?
  - [x] Listens and accepts clients
  - [x] Joins and spawns clients in minet
  - [x] Spawns vanilla clients

- [ ] Inventory transactions
- [ ] Movement physics - bedrock has a "Server Authoritative Movement" mode that the server can enable. If enabled, the client sends special movement packets where the server computes the physics and the player sends the inputs. Not exactly sure how this works in practice since the client still sends position packets. See https://github.com/Sandertv/gophertunnel/blob/05ac3f843dd60d48b9ca0ab275cda8d9e85d8c43/minecraft/protocol/packet/player_auth_input.go#L56

### CODE REVIEW

File: dist/datatypes/minecraft.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
+'use strict';
```
Comment (rom1504) [inline dist/datatypes/minecraft.js]: dist shouldn't be commited

**Comment by Karang** [comment]
Shouldn't you target the 1.1 branch to keep the diff clean ? Or merge it before ?

File: dist/datatypes/minecraft.js (hunk 1 above)
Comment (extremeheat) [reply dist/datatypes/minecraft.js]: Yeah there’s a lot of leftover code and dependencies that need to be removed. I think filfat was using Babel, it’s not required anymore

**Comment by rom1504** [comment]
I don't believe 1.1 branch was ever working, so targeting master makes sense.
We'll merge to master whenever it's clean and working enough (doesn't need to be perfect)

**Comment by rom1504** [comment]
github is saying there are merge conflict, can you rebase now to make things easier afterwards ? (just keep all your changes)

**Comment by extremeheat** [comment]
Fixed the merge issues. For the most part I’ve just added new files and not touched the old ones, I’ll clean it up at some point. 
Yeah, maybe it would’ve been better to merge 1.1 first, but it works for now I guess minus the big history. (If needed we can close this, merge 1.1 and reopen this)

Anyway, I looked into why clients couldn’t spawn, it looks like I set the raknet tps too low that the ACKs didn’t make it to the server in time so the server unnecessarily responds with old packets thinking they were lost. I workaround by sending ACKs faster, but it should be fixed in raknet properly so duplicate packets are properly handled. 

Also, the decryption checksum is currently broken. It can decrypt, but the checksum test seems to fail for some reason on vanilla. It’s not a big deal since the verification on the our side is optional (we only need to send a valid checksum to the other side)

**Comment by extremeheat** [comment]
Set raknet to run on its own worker, seems to be stable now and haven't had any issues. Also think I hunted down the issue with the checksum failing. Looks like it was indeed another issue with raknet...

**Comment by rom1504** [comment]
might be worth deleting babel and dist/ folder to simplify ?

**Comment by extremeheat** [comment]
I've cleaned up some of the junk, the rebase earlier failed because of annoying merge conflicts and would need resorting commits so I just did a normal merge there to fix the conflicts (didn't want to risk breaking history as some things are still useful references). 
Anyway, the world implementation is mostly done in bedrock-provider, only thing left to work on are tiles and entities. The caching implementation is complete in the provider also ([thanks mojang for actual docs](https://gist.github.com/Tomcc/4be79d3eafcd158c5059abd4ab2e8d35)!), just need to figure out a way to implement that into the protocol and how stuff using the library would work with it. 

Movement is still a WIP, need to implement some more packets for this to work properly I think. Aside from that, most of the hard stuff is complete. Also, I think it would be a good idea to rename this to bedrock-protocol or something like that for better discoverability/easier searching, I don't think there's anything to break. 

Also todo: 64-bit unsigned varints. Signed zigzag varints are pulled down from prismarine-nbt, but no support for 64-bit varints in protodef. We could either add that to the protocol here or put it in protodef.

**Comment by rom1504** [comment]
went through it, it looks fine
main comment is the data should go to minecraft data

about the name change, yeah let's do it

can you do a commit to rename all occurences ? I'll rename the repo

**Comment by rom1504** [comment]
if you can move the data to mcdata I think this is ready to merge. Would be good to do that so next changes will be easier to review

File: package.json [hunk 2]
```diff
@@ -1,29 +1,37 @@
...
+    "@xboxreplay/xboxlive-auth": "^3.3.3",
+    "aes-js": "^3.1.2",
+    "asn1": "^0.2.4",
+    "bedrock-provider": "github:extremeheat/bedrock-provider",
+    "debug": "^4.3.1",
+    "ec-pem": "^0.18.0",
+    "jsonwebtoken": "^8.5.1",
+    "jsp-raknet": "github:extremeheat/raknet#client",
+    "minecraft-folder-path": "^1.1.0",
+    "prismarine-nbt": "github:extremeheat/prismarine-nbt#le",
```
Comment (rom1504) [inline package.json]: this is released, you can use 1.5.0

File: package.json [hunk 3]
```diff
@@ -1,29 +1,37 @@
...
+    "protodef": "github:extremeheat/node-protodef#compiler",
```
Comment (rom1504) [inline package.json:29]: also released

File: package.json [hunk 4]
```diff
@@ -1,29 +1,37 @@
...
+    "@azure/msal-node": "^1.0.0-beta.6",
+    "@jsprismarine/jsbinaryutils": "^2.1.8",
+    "@xboxreplay/xboxlive-auth": "^3.3.3",
+    "aes-js": "^3.1.2",
+    "asn1": "^0.2.4",
+    "bedrock-provider": "github:extremeheat/bedrock-provider",
+    "debug": "^4.3.1",
+    "ec-pem": "^0.18.0",
+    "jsonwebtoken": "^8.5.1",
+    "jsp-raknet": "github:extremeheat/raknet#client",
```
Comment (rom1504) [inline package.json:26]: can you PR it and release it ?

File: package.json [hunk 5]
```diff
@@ -1,29 +1,37 @@
...
-    "prismarine-nbt": "^1.0.0",
-    "protodef": "^1.2.0",
-    "raknet": "^1.7.3",
-    "uuid-1345": "^1.0.1"
+    "@azure/msal-node": "^1.0.0-beta.6",
+    "@jsprismarine/jsbinaryutils": "^2.1.8",
+    "@xboxreplay/xboxlive-auth": "^3.3.3",
+    "aes-js": "^3.1.2",
+    "asn1": "^0.2.4",
+    "bedrock-provider": "github:extremeheat/bedrock-provider",
```
Comment (rom1504) [inline package.json:22]: same, please release

File: src/ConnWorker.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+const RakClient = require('jsp-raknet/client')
+const { Worker, isMainThread, parentPort } = require('worker_threads')
+const EncapsulatedPacket = require('jsp-raknet/protocol/encapsulated_packet')
```
Comment (rom1504) [inline src/ConnWorker.js]: this should be exposed

**Comment by extremeheat** [comment]
I’ll look at this tomorrow when I get the time. For moving the data to mcdata: the protocol json + some of the login packets go in mcdata and everything else stays here right?

Right now I generate the json from yaml and also precompile the js with protodef to make it easier to quickly debug. There’s some additional changes I made to protodef to expose the compiler thar to be PRed. 

Also I still need to add some samples and some  MITM code to figure out how the movements/inventory things work.

**Comment by rom1504** [comment]
Yeah indeed only protocol.json

Also could you delete yarn.lock ?

For the rest it's ok if it's done in other PRs.

**Review by rom1504 — APPROVED** [review]

**Comment by rom1504** [comment]
let's continue the work in small PRs on master

**Pull request merged**